Master CI failing since `810fe5d`: `1.6.13-SNAPSHOT` WALA artifacts unresolvable

## Summary

`master`'s build job has been failing since `810fe5d` (the bump of `org.codehaus.mojo:exec-maven-plugin` from 3.5.1 to 3.6.1, [#336](https://github.com/wala/ML/pull/336)). The failure is environmental, not caused by the bump itself — the cause is dependency-resolution failure for `com.ibm.wala:*:1.6.13-SNAPSHOT` artifacts.

## Failure Mode

```
[ERROR] Could not find artifact com.ibm.wala:com.ibm.wala.util:jar:1.6.13-SNAPSHOT
[ERROR] dependency: com.ibm.wala:com.ibm.wala.shrike:jar:1.6.13-SNAPSHOT (compile)
[ERROR] dependency: com.ibm.wala:com.ibm.wala.core:jar:1.6.13-SNAPSHOT (compile)
[ERROR] dependency: com.ibm.wala:com.ibm.wala.cast:jar:1.6.13-SNAPSHOT (compile)
[ERROR] dependency: com.ibm.wala:com.ibm.wala.cast.java:jar:1.6.13-SNAPSHOT (compile)
```

CI can't resolve the WALA SNAPSHOT artifacts. There's no accessible Maven repository serving them from this CI's perspective.

## Timeline (Build Job per Commit)

| SHA | Status | Description |
|---|---|---|
| `2dc3cb2` | green | Bump wala.version from 1.6.12-SNAPSHOT to **1.6.12** (#324) — last stable-version commit |
| `386a846` | green | "Use WALA SNAPSHOT version, upgrade to Java 25, ..." (#326) — switched back to `1.6.13-SNAPSHOT` |
| `35fb90b` | green | Add Copilot instructions (#330) |
| `ef22e49` | green | Remove duplicated build configurations (#328) |
| `55304c2` | green | Bump jacoco-maven-plugin (#337) |
| `fba6551` | green | Bump spotless-maven-plugin to 3.0.0 (#335) |
| `810fe5d` | **failure** (19 retries) | Bump exec-maven-plugin to 3.6.1 (#336) |

## Diagnosis

The SNAPSHOT switch in `386a846` introduced the latent fragility, but did *not* itself break the build — that commit and the next four built green, demonstrating the snapshots WERE resolvable at that time. The transition to red happened at `810fe5d`, but the bump there is unrelated to dependency resolution. Most likely:

- The WALA SNAPSHOT artifacts were rotated out / replaced / lost (typical Maven snapshot retention behavior), and earlier CI builds happened to hit the cache window where they were still resolvable.
- Or the snapshot repo URL/auth changed and only fresh resolution attempts surface the failure.
- Either way, `810fe5d` is the surfacing point, not the cause.

## Suggested Fixes

Two complementary directions:

1. **Pin to a stable WALA release** — `2dc3cb2` previously switched `1.6.12-SNAPSHOT` → `1.6.12` for exactly this fragility reason. The same logic suggests reverting the SNAPSHOT switch in `386a846` once `1.6.13` is published. If `1.6.13` isn't out yet, either pin to `1.6.12` again (with documented Java-version implications), or wait for the release.
2. **Add a snapshot repository** to `pom.xml`'s `<repositories>` section explicitly, pointing at a known-permanent location for WALA snapshots (e.g., Sonatype OSSRH snapshots repo if WALA publishes there). Documents the dependency on snapshot infrastructure and makes resolution reproducible.

## Impact

Master is unbuildable on CI. Open PRs (e.g., #432) inherit the failure regardless of their content, so they can't pass checks until master is restored. Automerge can't fire on those PRs.

## Related

- #336 — the exec-maven-plugin bump where the failure first surfaced (effect, not cause).
- #326 — the SNAPSHOT switch that introduced the latent fragility.
- #324 — earlier transition AWAY from SNAPSHOT (`1.6.12-SNAPSHOT` → `1.6.12`); same fragility motivated that.
- #432 — currently-open PR blocked by this issue.

## Tried Pinning To A Released WALA — Doesn't Work

Quick experiment off `upstream/master` to test the hypothesis that pinning to a released WALA version (instead of `1.6.13-SNAPSHOT`) would unblock CI:

- **Latest WALA release on Maven Central**: `1.6.10` (April 2025). There is no released `1.6.11`, `1.6.12`, or `1.6.13`.
- **Tried**: `<wala.version>1.6.13-SNAPSHOT</wala.version>` → `<wala.version>1.6.10</wala.version>`, `mvn -pl ml/com.ibm.wala.cast.python.ml -am compile`.
- **Result**: compile failure — wala/ML uses newer-than-1.6.10 WALA APIs:

  ```
  [ERROR] PythonSSAPropagationCallGraphBuilder.java:[834,9] cannot find symbol
    symbol:   method newFieldOperation(CGNode, int, int, boolean,
                  PythonConstraintVisitor.TrampolineWriter)
  [ERROR] PythonScopeMappingInstanceKeys.java:[57,3]
    method does not override or implement a method from a supertype
  ```

So the SNAPSHOT presence is intentional — wala/ML legitimately requires WALA APIs added after `1.6.10` (the trampoline-writer overload, plus the scope-mapping override-target). The fix shape isn't "bump to a release"; it's one of:

1. **Cut a WALA `1.6.11` (or higher) release** that includes those APIs, then pin wala/ML to it.
2. **Fix the SNAPSHOT publication pipeline** so `1.6.13-SNAPSHOT` resolves reliably from CI (the actual root cause of #433's symptoms — see [`810fe5d5`](https://github.com/wala/ML/commit/810fe5d5) timing).
3. **Vendor the needed WALA changes** into wala/ML's own classpath until upstream WALA cuts a release. (Heaviest option.)

(1) is cleanest if WALA's release cadence cooperates; (2) is the right long-term fix for SNAPSHOT consumers in general.

## Update: 1.7.1 And 1.6.12 Both Tried, Both Insufficient

Followed up on the released-WALA hypothesis with each of the post-1.6.10 releases (`v1.6.11`, `v1.6.12`, `v1.7.0`, `v1.7.1` per the [WALA release page](https://github.com/wala/WALA/releases)). Two distinct API-break boundaries straddle wala/ML:

### WALA 1.7.x (latest, including 1.7.1) — Compile Failure

`AstTranslator` interface changes:

```
[ERROR] PythonLoader.java:[384] incompatible types:
   Map<IBasicBlock, TypeReference[]>  vs.  Map<IBasicBlock, Set<TypeReference>>
[ERROR] PythonCAstToIRTranslator.java:[83] is not abstract and does not override
   doFieldWrite(WalkContext, int, CAstNode, CAstNode, CAstNode, int)
[ERROR] PythonCAstToIRTranslator.java:[200,298,336] method does not override or implement
   a method from a supertype
```

Adopting 1.7.x would require porting wala/ML to the new `AstTranslator` API surface (defineFunction signature, doFieldWrite override, etc.) — a real refactor, not a version bump.

### WALA 1.6.12 (last 1.6.x release) — Compile Clean, 72 Runtime Failures

`mvn -pl ml/com.ibm.wala.cast.python.ml.test -am test` against `<wala.version>1.6.12</wala.version>`:

- Builds cleanly.
- `TestTensorflow2Model` runs 353 tests (vs. 590 under `1.6.13-SNAPSHOT`) with **72 failures**.
- All 72 failures share the pattern `AssertionError: Function script ... must exist in call graph` — i.e., the call graph isn't picking up entrypoint methods that wala/ML's tests assume are present.
- This points at entrypoint/scope-mapping work that landed in WALA *between* 1.6.12 and 1.7.0 — exactly the contents of `1.6.13-SNAPSHOT`.

### Implication

`1.6.13-SNAPSHOT` isn't a stale pin to be bumped — it's the **only** WALA tip on which wala/ML's current code works correctly:

- 1.6.10/11/12: missing the entrypoint additions wala/ML's tests depend on.
- 1.7.x: the AstTranslator was refactored away from wala/ML's expectations.

So the path to unblocking CI isn't a version bump. The remaining options are:

1. **Get WALA to cut a `1.6.13` release** (or any tagged release containing the SNAPSHOT's contents) and pin to that. Cleanest if WALA's release cadence cooperates.
2. **Pin to a *dated* SNAPSHOT artifact** (e.g. `1.6.13-20250730.123456-N`) instead of the moving `-SNAPSHOT` tag. Avoids the "current SNAPSHOT failed to publish" race.
3. **Fix the SNAPSHOT publication pipeline** so `1.6.13-SNAPSHOT` resolves reliably from CI. Right long-term fix.
4. **Vendor the needed WALA changes** into wala/ML's classpath. Heaviest.

(2) is probably the fastest unblock, (1) the cleanest. Either way, the user-suggested "use a released version" doesn't directly work without one of the above.

## Correction To My Earlier Comment

In my [previous comment](https://github.com/wala/ML/issues/433#issuecomment-4348159561) I described `v1.6.13` (now released to Maven Central) as the durable fix for this issue. After actually running the experiment — bumping `<wala.version>1.6.13-SNAPSHOT</wala.version>` → `<wala.version>1.6.13</wala.version>`, porting test files, running the suite — I have to walk that back.

## Concrete Findings

1. **`v1.6.13` is on Maven Central**, at the requested SHA (`ea7ccf046`, parent `4eb96214e`). That part is correct.
2. **`wala/ML` upstream master does NOT compile against `v1.6.13` out of the box.** The catchTypes API change ([wala/WALA `f08697550`](https://github.com/wala/WALA/commit/f08697550), #1803) landed Dec 12, 2025 — well before the `v1.6.13` tag — and `wala/ML` upstream master still uses the pre-break `Map<…, TypeReference[]>` signature in `PythonLoader.java` and `PythonCAstToIRTranslator.java`. A port is needed (~30 lines of signature flips) plus the `Object[][]` → `List<TestCallGraphShape.GraphAssertion>` test-file port (~78 sites) plus an AssertJ runtime dep.
3. **After porting, the test suite still produces 370 tests / 72 failures.** All 72 are `TestTensorflow2Model.testModule*` failing with `AssertionError: Function script ... must exist in call graph`. **Same 72 failures reproduce against the locally-built `1.6.13-SNAPSHOT` from a much-newer WALA tip** — so they aren't caused by `v1.6.13` specifically; they're a pre-existing condition on upstream `wala/ML` that exists across multiple WALA tips.
4. **Last green `wala/ML` CI run was Feb 3, 2026** ([run 21616994868](https://github.com/wala/ML/actions/runs/21616994868)). By then WALA master had had the catchTypes break for ~7 weeks. How CI was passing during that window is an open puzzle I didn't trace — could be CI caching an older WALA jar, could be a different `wala/ML` code state I missed, could be `--remote` not actually picking up the break for some reason.

## What This Means

- The CI workflow PR ([#443](https://github.com/wala/ML/pull/443)) — drop `--remote` — is still the cleanest immediate unblock. It pins to the committed `c0e260a62` submodule SHA (Sep 22, 2025, before the catchTypes break), so the existing `PythonLoader`/`PythonCAstToIRTranslator` code stays compatible. The 72 `testModule*` failures still need investigation, but they're a separate issue from the SNAPSHOT-resolution race that #443 closes.
- `v1.6.13`'s real value is narrower than I claimed: it's a stable, named WALA coordinate for downstream consumers who want to pin to a release, but adopting it inside `wala/ML` requires the AstTranslator port + the GraphAssertion port + investigating the 72 testModule failures.
- The JEP migration ([#440](https://github.com/wala/ML/issues/440), [#441](https://github.com/wala/ML/issues/441)) remains the durable answer — eliminates the Jython/AstTranslator coupling entirely.

Sorry for the noise on the earlier framing.

--- a/.github/workflows/continuous-integration.yml
+++ b/.github/workflows/continuous-integration.yml
@@ -15,8 +15,6 @@ jobs:
       uses: actions/checkout@v5
       with:
         submodules: 'recursive'
-    - name: Update the WALA submodule with its latest commit
-      run: git submodule update --recursive --remote
     - name: Set up JDK
       uses: actions/setup-java@v4
       with:

